crates/shared/src/price_estimation/buffered.rs

```diff
@@ -0,0 +1,248 @@
+                .shared()
+        };
+
+        // Create a `SharedEstimationRequest` for each individual `Query` of the batch. This
+        // makes it possible for a `batch_2` to await the queries which it is interested in of the
+        // in-flight `batch_1`. Even if the estimator which requested `batch_1` stops polling it, the
+        // estimator of `batch_2` can still poll `batch_1` to completion by polling the
+        // `SharedEstimationRequest` it is actually interested in.
+        //
+        // Build those shared futures up front to keep the critical section short when inserting
+        // them into the `active_requests`.
+        #[allow(clippy::needless_collect)]
+        let individual_requests_for_batch: Vec<_> = remaining_queries
+            .iter()
+            .flatten()
+            .enumerate()
+            .map(|(index, query)| {
+                let fetch_remaining_estimates = fetch_remaining_estimates.clone();
+                (
+                    *query,
+                    async move { fetch_remaining_estimates.await[index].clone() }
+                        .boxed()
+                        .shared(),
+                )
+            })
+            .collect();
+        self.inner
+            .in_flight_requests
+            .lock()
+            .unwrap()
+            // It is possible that someone stored a `SharedEstimationRequest` in the meantime which
+            // could be overwritten now. In those rare cases 2 identical estimation requests would
+            // be in-flight at the same time but it wouldn't produce errors. No memory would leak
+            // and a calling estimator could still decide to stop polling a future while other
+            // estimators interested in the result are able poll the future to completion.
+            .extend(individual_requests_for_batch.into_iter());
+
+        // Await all the estimates we need (in-flight and the new ones) in parallel.
+        let results = futures::join!(
+            futures::future::join_all(
+                active_requests
+                    .iter()
+                    .flatten()
+                    .cloned()
+                    .collect::<Vec<_>>(),
+            ),
+            fetch_remaining_estimates
+        );
+        let (mut in_flight_results, mut new_results) =
+            (results.0.into_iter(), results.1.into_iter());
+
+        // Return the results of new and in-flight requests merged into one.
+        active_requests
+            .iter()
+            .map(|request| match request {
+                Some(_) => in_flight_results.next().unwrap(),
+                None => new_results.next().unwrap(),
+            })
+            .collect()
+    }
+}
+
+#[async_trait::async_trait]
+impl PriceEstimating for BufferingPriceEstimator {
+    async fn estimates(&self, queries: &[Query]) -> Vec<EstimationResult> {
+        let result = self.estimate_buffered(queries).await;
+        self.inner.collect_garbage();
```

Comment:
Cleaning up only after completed estimates a problem. Someone could keep creating new estimates which internally creates in flight futures and then dropping those estimate futures before this call to estimate_buffered completes. This causes collect_garbage to never run.
There are several solutions:

1. Ignore the problem because it is a rare edge case and we complete these futures most of the time
2. Create background clean up task
3. Create a drop guard that performs the cleanup.

3 Would look like
```rust
        struct DropGuard<'a>(&'a BufferingPriceEstimator);
        impl<'a> Drop for DropGuard<'a> {
            fn drop(&mut self) {
                self.0.inner.collect_garbage();
            }
        }
        let _guard = DropGuard(self);
        let result = self.estimate_buffered(queries).await;
        result
```

I'm not sure what is best.

Reply:
With a drop guard we could even use the still existing `&[Queries]` to only garbage collect the exact queries we might have created.

Reply:
Instead of a drop guard, shouldn't we just garbage collect before calling `estimate_buffered`? This way we remove any shared futures that have already resolved or are no longer being polled (so we don't use stale requests).

This way we don't have to change the code to use `WeakShared` either.

Reply:
I think the drop guard should be the way to go.
Coming to think of it the current implementation has a problem:
Imagine `batch_1` estimates many queries. This will cause us to store many `Shared` (or `WeakShared`) futures.
Immediately afterwards a totally unrelated batch finishes and collects garbage. This would cause many futures to be dropped which are currently unused but could have gotten used until `batch_a` terminates. This would cause us to issue many more requests than needed.

Reply:
That does make me think of a new problem. If the Shared has internally finished but someone keeps a reference to a future using alive because for whatever reason they don't poll it, then we continue to give out more copies of this Shared even though the internal future might be very old.
I don't immediately see how to fix that with this design.

Reply:
Another alternative design idea I had but that I didn't think the implications through is to `tokio::spawn` the inner futures. That might simplify some things or give more control in some  cases.